atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/service/DefaultKnownCommittedTimestamps.java

```diff
@@ -0,0 +1,173 @@
+
+    // Naughty way of returning 2 values!
+    private synchronized byte[] ensureMapLoadedFromDatabase() {
+        try {
+            byte[] value = databaseMapLoader.get();
+            knownCommittedTimestamps = OBJECT_MAPPER.readValue(value, SerializableTimestampSet.class);
+            return value;
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private byte[] loadMapFromDatabase() {
+        Map<Cell, Value> dbRead = keyValueService.get(
+                TransactionConstants.KNOWN_COMMITTED_TIMESTAMPS,
+                ImmutableMap.of(MAGIC_CELL, AtlasDbConstants.TRANSACTION_TS));
+        return dbRead.get(MAGIC_CELL).getContents();
+    }
+
+    private boolean shouldLoadMapFromDatabase(long timestamp) {
+        if (knownCommittedTimestamps.longRanges().isEmpty()) {
+            return true;
+        }
+
+        // Bad Demeter. This logic should be in SerializableTimestampSet in the prod version
+        Range<Long> highestRange = knownCommittedTimestamps
+                .rangeSetView()
+                .asDescendingSetOfRanges()
+                .iterator()
+                .next();
+        Preconditions.checkState(
+                highestRange.hasUpperBound(), "Infinite ranges not allowed in KnownCommittedTimestamps");
+        long maxBound = highestRange.upperBoundType() == BoundType.CLOSED
+                ? highestRange.upperEndpoint()
+                : highestRange.upperEndpoint() - 1;
+        return timestamp > maxBound;
+    }
+
+    @Override
+    public void addInterval(Range<Long> knownCommittedInterval) {
+        WriteLock writeLock = lock.writeLock();
+        writeLock.lock();
+        try {
+            // update the DB first and then update local state
+            // if DB update fails, local state is not updated
+            // DB may be ahead of local state, that is OK
+            while (true) {
+                byte[] oldValue = ensureMapLoadedFromDatabase();
+
+                if (knownCommittedTimestamps.rangeSetView().encloses(knownCommittedInterval)) {
+                    return;
+                }
+
+                // This is responsible for coalescing
+                SerializableTimestampSet withNewInterval =
+                        SerializableTimestampSet.copyWithRange(knownCommittedTimestamps, knownCommittedInterval);
+                byte[] newValue;
+                try {
+                    newValue = OBJECT_MAPPER.writeValueAsBytes(withNewInterval);
+                } catch (JsonProcessingException e) {
+                    throw new RuntimeException(e);
+                }
+
+                try {
+                    if (knownCommittedTimestamps.longRanges().isEmpty()) {
+                        // Put new cell in
+                        keyValueService.checkAndSet(CheckAndSetRequest.newCell(
+                                TransactionConstants.KNOWN_COMMITTED_TIMESTAMPS, MAGIC_CELL, newValue));
+                    } else {
```

Comment:
The whole set of ranges is saved in one cell. i wonder how confident we are that we will not have to modify this algorithm soon enough due to the size of this set. Is there a way to gain visibility on how big this set could be for internal atlas proxy?